packages/components/src/components/files/WebDav.ts

```diff
@@ -0,0 +1,151 @@
+        }) as WebDavFile;
+    }
+}
+
+function getWebDavUrl(containerPath: string, directory?: string, createIntermediates?: boolean) {
+    let url =
+        ActionURL.getContextPath() +
+        '/_webdav' +
+        ActionURL.encodePath(containerPath) +
+        '/' +
+        encodeURIComponent('@files');
+
+    if (directory) {
+        url += '/' + directory;
+    }
+
+    if (createIntermediates) {
+        url += '?createIntermediates=' + createIntermediates;
+    }
+
+    return url;
+}
+
+export function getWebDavFiles(
+    containerPath: string,
+    directory?: string,
+    includeDirectories?: boolean
+): Promise<Map<string, WebDavFile>> {
+    return new Promise((resolve, reject) => {
+        const url = getWebDavUrl(containerPath, directory);
+
+        return Ajax.request({
+            url: url + '?method=JSON',
+            method: 'GET',
+            success: Utils.getCallbackWrapper(response => {
+                // Filter directories and create webdav files
+                const filteredFiles = response.files.reduce((filtered, file) => {
+                    if (includeDirectories || !file.collection) {
+                        return filtered.set(file.text, WebDavFile.create(file));
+                    }
+                }, Map<string, WebDavFile>());
+
+                resolve(filteredFiles);
+            }),
+            failure: Utils.getCallbackWrapper(
+                response => {
+                    console.error('Problem retrieving webDav files for container ' + containerPath);
+                    reject(response);
+                },
+                null,
+                false
+            ),
+        });
+    });
+}
+
+export function uploadWebDavFile(
+    file: File,
+    containerPath: string,
+    directory?: string,
+    createIntermediates?: boolean
+): Promise<string> {
+    return new Promise((resolve, reject) => {
+        const form = new FormData();
+        form.append('file', file);
+
+        const url = getWebDavUrl(containerPath, directory, createIntermediates);
+
+        Ajax.request({
+            url,
+            method: 'POST',
+            form,
+            success: Utils.getCallbackWrapper(() => {
+                resolve(file.name);
+            }),
+            failure: Utils.getCallbackWrapper(
+                () => {
+                    console.error('failure uploading file ' + file.name);
+                    reject(file.name);
+                },
+                null,
+                false
+            ),
+        });
+    });
+}
+
+export function uploadWebDavFiles(
+    files: Map<string, File>,
+    containerPath: string,
+    directory?: string,
+    createIntermediates?: boolean,
+    onSuccess?: () => any,
+    onFailure?: (any) => any
+): void {
+    if (files.size === 0) {
+        return;
+    }
+
+    // DavController does not support multiple file uploads, so we do one at a time.
+    const promises = files.reduce((uploads, file) => {
```

Comment:
This should be `.map` instead of `.reduce` then you don't need to worry about modifying the array at all, you can just return a new uploadWebDavFile promise:

```
const promises = files.map((uploads, file) => {
    return uploadWebDavFile(file, containerPath, directory, createIntermediates);
});
```

Reply:
.map on a Map returns another Map, then you have to get into some other weird conversion of Map to array